src/serialization/json_object.cpp

```diff
@@ -0,0 +1,1124 @@
+// Copyright (c) 2016-2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "json_object.h"
+
+#include "string_tools.h"
+
+namespace cryptonote
+{
+
+namespace json
+{
+
+void toJsonValue(rapidjson::Document& doc, const std::string& i, rapidjson::Value& val)
+{
+  val = rapidjson::Value(i.c_str(), doc.GetAllocator());
+}
+
+void fromJsonValue(const rapidjson::Value& val, std::string& str)
+{
+  if (!val.IsString())
+  {
+    throw WRONG_TYPE("string");
+  }
+
+  str = val.GetString();
+}
+
+void toJsonValue(rapidjson::Document& doc, bool i, rapidjson::Value& val)
+{
+  val.SetBool(i);
+}
+
+void fromJsonValue(const rapidjson::Value& val, bool& b)
+{
+  if (!val.IsBool())
+  {
+    throw WRONG_TYPE("boolean");
+  }
+  b = val.GetBool();
+}
+
+void toJsonValue(rapidjson::Document& doc, const uint8_t& i, rapidjson::Value& val)
+{
+  val = rapidjson::Value(i);
+}
+
+
+void fromJsonValue(const rapidjson::Value& val, uint8_t& i)
+{
+  if (!val.IsUint())
+  {
+    throw WRONG_TYPE("unsigned integer");
+  }
+
+  if (val.GetUint() > 0xFF)
+  {
+    throw WRONG_TYPE("type too large");
+  }
+
+  i = (uint8_t)( val.GetUint() & 0xFF);
+}
+
+void toJsonValue(rapidjson::Document& doc, const int8_t& i, rapidjson::Value& val)
+{
+  val = rapidjson::Value(i);
+}
+
+
+void fromJsonValue(const rapidjson::Value& val, int8_t& i)
+{
+  if (!val.IsInt())
+  {
+    throw WRONG_TYPE("integer");
+  }
+
+  if (val.GetInt() > 0xFF)
```

Comment:
I re-worked this logic in a PR issued to @tewinget repo/branch. The limits are now checked using `numeric_limits<...>min/max`.